ext/standard/array.c

```diff
@@ -91,6 +91,11 @@ PHP_MSHUTDOWN_FUNCTION(array) /* {{{ */
 }
 /* }}} */
 
+static zend_always_inline bool zend_may_modify_array_in_place(const zval *arg)
+{
+	return !(GC_FLAGS(Z_ARRVAL_P(arg)) & (IS_ARRAY_IMMUTABLE | IS_ARRAY_PERSISTENT)) && Z_REFCOUNT_P(arg) == 1;
+}
+
```

Comment:
It makes sense to add assertion ``ZEND_ASSERT(Z_TYPE_P(arg) == IS_ARRAY)``

Maybe it makes sense to make this function even more general.
```c
static zend_always_inline bool zend_may_modify_in_place(const zval *arg)
{
	return Z_REFCOUNTED_P(arg) && !(GC_FLAGS(Z_ARRVAL_P(arg)) & (GC_IMMUTABLE | GC_PERSISTENT)) && Z_REFCOUNT_P(arg) == 1;
}
```

This may be moved to ``zend_types.h``